static/src/javascripts/projects/common/modules/commercial/dfp/messenger.js

```diff
@@ -0,0 +1,149 @@
+            return;
+        }
+
+        try {
+            // Even though the postMessage API allows passing objects as-is, the
+            // serialisation/deserialisation is slower than using JSON
+            // Source: https://bugs.chromium.org/p/chromium/issues/detail?id=536620#c11
+            var data = JSON.parse(event.data);
+        } catch( ex ) {
+            respond(error418, null);
+            return;
+        }
+
+        if (!isValidPayload(data)) {
+            respond(error400, null);
+            return;
+        }
+
+        // If there is no routine attached to this event type, we just answer
+        // with an error code
+        if (!listeners[data.type].length) {
+            respond(formatError(error405, data.type), null);
+            return;
+        }
+
+        var promise = listeners[data.type]
+        // We offer, but don't impose, the possibility that a listener returns
+        // a value that must be sent back to the calling frame. To do this,
+        // we pass the cumulated returned value as a second argument to each
+        // listener. Notice we don't try some clever way to compose the result
+        // value ourselves, this would only make the solution more complex.
+        // That means a listener can ignore the cumulated return value and
+        // return something else entirely—life is unfair.
+        .map(function (listener) {
+            return function promiseCallback(ret) {
+                var thisRet = listener(data.value, ret);
+                return thisRet === undefined ? ret : thisRet;
+            };
+        })
+        // We don't know what each callack will be made of, we don't want to.
+        // And so we wrap each call in a promise chain, in case one drops the
+        // occasional fastdom bomb in the middle.
+        .reduce(function (promise, promiseCallback) {
+            return promise.then(promiseCallback);
+        }, Promise.resolve(true));
+
+        return promise.then(function (response) {
+            respond(null, response);
+        }).catch(function (ex) {
+            reportError(ex, { feature: 'native-ads' });
+            respond(formatError(error500, ex), null);
+        });
+
+        function respond(error, result) {
+            event.source.postMessage(JSON.stringify({ error: error, result: result }), currentHost);
+        }
+    }
+
+    // Until DFP provides a way for us to identify with 100% certainty our
+    // in-house creatives, we are left with doing some basic tests
+    // such as validating the anatomy of the payload and whitelisting
+    // event type
+    function isValidPayload(payload) {
+        return 'type' in payload &&
+            'value' in payload &&
+            payload.type in listeners;
+    }
+
+    // Cheap string formatting function. It accepts as its first argument
+    // a string where each occurence of %% will be replaced by the following
+    // arguments
+    function formatError() {
```

Comment:
Kinda complex for what it does (your errors have only one `%%`).

Tiny correction on your comment, what you say is true if your String.replace is called with a string. Works differently with a regex.

For instance you could write

```js
function formatError () {
    if (arguments.length < 2) {
        return arguments[0] || '';
    }

    var replacements = [].slice.call(arguments, 1);
    return arguments[0].replace(/%%/g, function () {
        return replacements.shift();
    });
}
```
Not that it's much easier to read